### ISSUE TITLE
Issue when assigning values to nested fields

### ISSUE THREAD

**Issue by ruizmurcia** [Automattic/mongoose#9459]
**Do you want to request a *feature* or report a *bug*?**
Report a bug

**What is the current behavior?**
From v5.9.10 and later we have experienced a change in the behaviour when assigning a value to a nested property of a model. Let's take this script as an example:

```javascript
'use strict';

const mongoose = require('mongoose');

const { Schema } = mongoose;

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test', {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });

  await mongoose.connection.dropDatabase();

  const preferencesSchema = mongoose.Schema({
    notifications:  {
      email: Boolean,
      push: Boolean
    },
    keepSession: Boolean
  },{ _id: false });

  const User = mongoose.model('User', Schema({
    email: String,
    username: String,
    preferences: preferencesSchema
  }));

  const userFixture = {
    email: 'foo@bar.com',
    username: 'foobars',
    preferences: {
      keepSession: true,
      notifications: {
        email: false,
        push: false
      }
    }
  };

  const userWithEmailNotifications = Object.assign({}, userFixture, {
    'preferences.notifications': { email: true}
  })

  const testUser = new User(userWithEmailNotifications)

  console.log('User', testUser);
}
```
The result is that no change is made to the nested property:
```
User {
  _id: 5f76e44107612fccf0b2f733,
  email: 'foo@bar.com',
  username: 'foobars',
  preferences: { keepSession: true, notifications: { email: false, push: false } }
}
```

**What is the expected behavior?**
In v5.9.9 and previous versions the same code results in changing the property:
```
User {
  _id: 5f76e4b01dc7eacd204ab418,
  email: 'foo@bar.com',
  username: 'foobars',
  preferences: { keepSession: true, notifications: { email: true, push: false } }
}
```
After looking for differences between v5.9.9 and v5.9.10 we have found that the issue is cause by the change in:
https://github.com/Automattic/mongoose/commit/8fea1d9121b5e87eb66d45b667fd2bd5baed0f57

This feels like a bug, though it may be intended behavior.

Could you please give us some clarification on the issue?

Thank you so much!

**Comment by vkarpov15** [comment]
Thank you for looking into this. This issue is due to Mongoose not handling nested paths within nested subdocs when setting a dotted path correctly, in either the v5.9.9 or the v5.10.10 behavior. The correct output is:

```javascript
User {
  _id: 5f76e4b01dc7eacd204ab418,
  email: 'foo@bar.com',
  username: 'foobars',
  preferences: { keepSession: true, notifications: { email: true } }
}
```

Because you're explicitly setting `preferences.notifications`, as opposed to just `preferences.notifications.email`. The v5.9.9 behavior was due to the fact that Mongoose wasn't able to find the correct schema type for `preferences.notification` when setting.

In v5.10.8 we'll change it so that the output is `notifications: { email: true }`. Does that work for you or do you need a workaround?

**Comment by ruizmurcia** [comment]
Thank you very much quick reply.

In our test suit we have a base object that we modified in each test depending on the conditions. What we expect is to merge the existing properties of the object `userFixture`, with the new properties that we have specified using the `Object.assign` method. We don't want to just overwrite them.

The behaviour in version v5.9.9 and previous was perfect for us, and we expect to get the following:

```
User {
  _id: 5f76e4b01dc7eacd204ab418,
  email: 'foo@bar.com',
  username: 'foobars',
  preferences: { keepSession: true, notifications: { email: true, push: false } }
}
```

Is there any workaround that we can implement to have this same result even in v5.10.8?
